File: bm_red_black_tree.rb
$:.unshift(File.expand_path(File.dirname(__FILE__) + "/../../lib"))

require 'red_black_tree'

def rbt_bm
  n = 100000
  a1 = []; n.times { a1 << rand(999_999) }
  a2 = []; n.times { a2 << rand(999_999) }
  start = Time.now

  n = 100_000

  tree = RedBlackTree.new

  n.times {|i| tree.add(i) }
  n.times { tree.delete(tree.root) }

  tree = RedBlackTree.new
  a1.each {|e| tree.add(e) }
  a2.each {|e| tree.search(e) }
  tree.inorder_walk {|key| key + 1 }
  tree.reverse_inorder_walk {|key| key + 1 }
  n.times { tree.minimum }
  n.times { tree.maximum }

  return Time.now - start
end

N = (ARGV[0] || 5).to_i

N.times do
  puts rbt_bm
end

File: bm_red_black_tree.py
import sys
import time
import random
from red_black_tree import RedBlackTree
import sys

minor = sys.version_info[0]
if minor >= 3:
  xrange = range

def rbt_bm():
  start = time.time()
  n = 100000

  tree = RedBlackTree()

  for i in xrange(0, n):
    tree.add(i)
  for i in xrange(0, n):
    tree.delete(tree.root)

  tree = RedBlackTree()
  for i in xrange(0, n):
    tree.add(random.randint(0, 999999))
  for i in xrange(0, n):
    tree.search(random.randint(0, 999999))
  for key in tree.inorder_walk():
    key + 1
  for key in tree.reverse_inorder_walk():
    key + 1
  for i in xrange(0, n):
    tree.maximum()
  for i in xrange(0, n):
    tree.minimum()

  return time.time() - start

if len(sys.argv) > 1:
  n = int(sys.argv[1])
else:
  n = 5

for i in xrange(0, n):
  print("%02f" % rbt_bm())
